control/noise/handshake.go

```diff
@@ -0,0 +1,430 @@
+	"golang.org/x/crypto/curve25519"
+	"golang.org/x/crypto/hkdf"
+	"golang.org/x/crypto/poly1305"
+	"tailscale.com/types/key"
+)
+
+const (
+	// protocolName is the name of the specific instantiation of the
+	// Noise protocol we're using. Each field is defined in the Noise
+	// spec, and shouldn't be changed unless we're switching to a
+	// different Noise protocol instance.
+	protocolName = "Noise_IK_25519_ChaChaPoly_BLAKE2s"
+	// protocolVersion is the version of the Tailscale base
+	// protocol that Client will use when initiating a handshake.
+	protocolVersion = 1
+	// protocolVersionPrefix is the name portion of the protocol
+	// name+version string that gets mixed into the Noise handshake as
+	// a prologue.
+	//
+	// This mixing verifies that both clients agree that
+	// they're executing the Tailscale control protocol at a specific
+	// version that matches the advertised version in the cleartext
+	// packet header.
+	protocolVersionPrefix = "Tailscale Control Protocol v"
+	invalidNonce          = ^uint64(0)
+)
+
+func protocolVersionPrologue(version int) []byte {
+	ret := make([]byte, 0, len(protocolVersionPrefix)+5) // 5 bytes is enough to encode all possible version numbers.
+	ret = append(ret, protocolVersionPrefix...)
+	return strconv.AppendUint(ret, uint64(version), 10)
+}
+
+// Client initiates a Noise client handshake, returning the resulting
+// Noise connection.
+//
+// The context deadline, if any, covers the entire handshaking
+// process.
+func Client(ctx context.Context, conn net.Conn, machineKey key.Private, controlKey key.Public) (*Conn, error) {
+	if deadline, ok := ctx.Deadline(); ok {
+		if err := conn.SetDeadline(deadline); err != nil {
+			return nil, fmt.Errorf("setting conn deadline: %w", err)
+		}
+		defer func() {
+			conn.SetDeadline(time.Time{})
+		}()
+	}
+
+	var s symmetricState
+	s.Initialize()
+
+	// prologue
+	s.MixHash(protocolVersionPrologue(protocolVersion))
+
+	// <- s
+	// ...
+	s.MixHash(controlKey[:])
+
+	// -> e, es, s, ss
+	init := mkInitiationMessage()
+	machineEphemeral := key.NewPrivate()
+	machineEphemeralPub := machineEphemeral.Public()
+	copy(init.EphemeralPub(), machineEphemeralPub[:])
+	s.MixHash(machineEphemeralPub[:])
+	if err := s.MixDH(machineEphemeral, controlKey); err != nil {
+		return nil, fmt.Errorf("computing es: %w", err)
+	}
+	machineKeyPub := machineKey.Public()
+	s.EncryptAndHash(init.MachinePub(), machineKeyPub[:])
+	if err := s.MixDH(machineKey, controlKey); err != nil {
+		return nil, fmt.Errorf("computing ss: %w", err)
+	}
+	s.EncryptAndHash(init.Tag(), nil) // empty message payload
+
+	if _, err := conn.Write(init[:]); err != nil {
+		return nil, fmt.Errorf("writing initiation: %w", err)
+	}
+
+	// Read in the payload and look for errors/protocol violations from the server.
+	var resp responseMessage
+	if _, err := io.ReadFull(conn, resp.Header()); err != nil {
+		return nil, fmt.Errorf("reading response header: %w", err)
+	}
+	if resp.Version() != protocolVersion {
```

Comment:
I did not specify the control's message to have a version field because the version is not negotiated: the client offers only one, and if the server doesn't like it it will send an error.

Having it on the wire is just another thing to check.